### PR TITLE
Add BWC for retention leases

### PR DESCRIPTION
**Pull request by jasontedor**
We have to handle the case of a < 6.7.0 and >= 6.7.0 mixed cluster wherein a primary on >= 6.7.0 would otherwise send retention leases to a < 6.7.0 node which would not understand them. This commit adds BWC for this case, and adds a test to ensure that we behave properly here.

Relates #37165

### CODE REVIEW

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: qa/retention-lease-bwc/src/test/java/org/elasticsearch/add_retention_lease/RetentionLeaseBwcIT.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+
+public class RetentionLeaseBwcIT extends ESRestTestCase {
+
+    @SuppressForbidden(reason = "debug")
```
Comment (jasontedor) [inline qa/retention-lease-bwc/src/test/java/org/elasticsearch/add_retention_lease/RetentionLeaseBwcIT.java]:
```suggestion
```

File: qa/retention-lease-bwc/src/test/java/org/elasticsearch/add_retention_lease/RetentionLeaseBwcIT.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+package org.elasticsearch.add_retention_lease;
+
+import org.apache.http.HttpHost;
+import org.elasticsearch.Version;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.SuppressForbidden;
```
Comment (jasontedor) [inline qa/retention-lease-bwc/src/test/java/org/elasticsearch/add_retention_lease/RetentionLeaseBwcIT.java]:
```suggestion
```

File: qa/retention-lease-bwc/src/main/java/org/elasticsearch/retention_lease_bwc/RestAddRetentionLeaseHandler.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                                new RestActionListener<Collection<RetentionLeaseActions.Response>>(channel) {
+
+                                    @Override
+                                    protected void processResponse(
+                                            final Collection<RetentionLeaseActions.Response> responses) throws Exception {
+                                        final XContentBuilder builder = channel.newBuilder().startObject().endObject();
+                                        channel.sendResponse(new BytesRestResponse(RestStatus.OK, builder));
+                                    }
+
+                                }, numberOfShards, Collections.emptyList());
```
Comment (jasontedor) [inline qa/retention-lease-bwc/src/main/java/org/elasticsearch/retention_lease_bwc/RestAddRetentionLeaseHandler.java]:
```suggestion
                                },
                                numberOfShards,
                                Collections.emptyList());
```

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/docbldesx

**Review by dnhatn — APPROVED** [review]
I am not sure if it's worth having a separate QA module for this BWC change. But I am fine with this PR since it will go to 6.7 only. Thanks @jasontedor.

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by jasontedor** [comment]
Thanks for reviewing @dnhatn and @martijnvg.

**Pull request merged**